File: Math challenge/Challanges/SPHAIR.m
//
//  SPHAIR.m
//  Math challenge
//
//  Created by Lukas Bischof on 09.01.18.
//  Copyright © 2018 Lukas Bischof. All rights reserved.
//

#import "SPHAIR.h"

@implementation SPHAIR

- (Challange *)next
{
    NSInteger result = randr(1, 100);
    NSInteger denominator = randr(2, 9);
    
    NSMutableArray *possibleNumeratorValues = [@[@2, @3, @4, @5, @6, @7, @8, @9] mutableCopy];
    NSUInteger len = possibleNumeratorValues.count;
    
    NSInteger x = result * denominator;
    BOOL hasChallange = NO;
    NSInteger numerator = 0;
    for (short i = 0; i < len; i++) {
        NSUInteger indx = randr(0, possibleNumeratorValues.count - 1);
        NSInteger randNumerator = [possibleNumeratorValues[indx] integerValue];
        double tempResult = (double)x / (double)randNumerator;
        
        if (tempResult == (double)floorf(tempResult) && randNumerator != denominator) {
            numerator = randNumerator;
            hasChallange = YES;
            break;
        } else {
            [possibleNumeratorValues removeObjectAtIndex:indx];
        }
    }
    
    if (hasChallange) {
        NSInteger multiplicator = x / numerator;
        return [[Challange alloc] initWithText:SWF(@"%ld/%ld * %ld", (long)numerator, (long)denominator, (long)multiplicator) andAnswer:result];
    } else {
        return [self next];
    }
}

- (NSString *)challangeDescription
{
    return @"SPHAIR";
}

@end

File: Math challenge/challenges/SPHAIR.swift
//
//  SPHAIR.swift
//  Math Challenge
//
//  Created by Lukas Bischof on 24.10.19.
//  Copyright © 2019 Lukas Bischof. All rights reserved.
//

class SPHAIR: MathChallengeCategory {
  let challengeDescription: String = "SPHAIR"

  func next() -> Challenge {
    let result = randomInt(1, 100)
    let denominator = randomInt(2, 9)

    var possibleNumeratorValues = [2, 3, 4, 5, 6, 7, 8, 9]
    let len = possibleNumeratorValues.count

    let x = result * denominator
    var hasChallenge = false
    var numerator = 0

    for _ in 0..<len {
      let indx = randomInt(0, possibleNumeratorValues.count - 1)
      let randNumerator = possibleNumeratorValues[indx]
      let tempResult = Double(x) / Double(randNumerator)

      if tempResult == Double(tempResult.rounded(.down)) && randNumerator != denominator {
        numerator = randNumerator
        hasChallenge = true
        break
      } else {
        possibleNumeratorValues.remove(at: indx)
      }
    }

    if hasChallenge {
      let multiplication = x / numerator
      return generateChallenge(numerator, denominator, multiplication, result)
    } else {
      return next()
    }
  }

  private func generateChallenge(_ numerator: Int, _ denominator: Int, _ multiplication: Int, _ result: Int) -> Challenge {
    let latex = "\\frac{\(numerator)}{\(denominator)} * \(multiplication)"
    return Challenge(text: "\(numerator)/\(denominator) * \(multiplication)", latex: latex, andAnswer: result)
  }
}
